File: src/main/java/net/minestom/server/utils/collection/MappedCollection.java
package net.minestom.server.utils.collection;

import org.jetbrains.annotations.ApiStatus;

import java.util.Collection;
import java.util.Iterator;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;

@ApiStatus.Internal
public record MappedCollection<O, R>(Collection<O> original,
                                     Function<O, R> mapper) implements Collection<R> {
    public static <O extends AtomicReference<R>, R> MappedCollection<O, R> plainReferences(Collection<O> original) {
        return new MappedCollection<>(original, AtomicReference::getPlain);
    }

    @Override
    public int size() {
        return original.size();
    }

    @Override
    public boolean isEmpty() {
        return original.isEmpty();
    }

    @Override
    public boolean contains(Object o) {
        for (var entry : original) {
            if (mapper.apply(entry).equals(o)) return true;
        }
        return false;
    }

    @Override
    public Iterator<R> iterator() {
        var iterator = original.iterator();
        return new Iterator<>() {
            @Override
            public boolean hasNext() {
                return iterator.hasNext();
            }

            @Override
            public R next() {
                return mapper.apply(iterator.next());
            }
        };
    }

    @Override
    public Object [] toArray() {
        // TODO
        throw new UnsupportedOperationException("Unsupported array object");
    }

    @Override
    public <T> T [] toArray(T [] a) {
        // TODO
        throw new UnsupportedOperationException("Unsupported array generic");
    }

    @Override
    public boolean containsAll(Collection<?> c) {
        if (c.size() > original.size()) return false;
        for (var entry : c) {
            if (!contains(entry)) return false;
        }
        return true;
    }

    @Override
    public boolean add(R t) {
        throw new UnsupportedOperationException("Unmodifiable collection");
    }

    @Override
    public boolean remove(Object o) {
        throw new UnsupportedOperationException("Unmodifiable collection");
    }

    @Override
    public boolean addAll(Collection<? extends R> c) {
        throw new UnsupportedOperationException("Unmodifiable collection");
    }

    @Override
    public boolean removeAll(Collection<?> c) {
        throw new UnsupportedOperationException("Unmodifiable collection");
    }

    @Override
    public boolean retainAll(Collection<?> c) {
        throw new UnsupportedOperationException("Unmodifiable collection");
    }

    @Override
    public void clear() {
        throw new UnsupportedOperationException("Unmodifiable collection");
    }
}

File: src/main/java/net/minestom/server/utils/collection/MappedCollection.kt
package net.minestom.server.utils.collection

import java.util.AbstractMap.SimpleImmutableEntry
import java.util.function.BiConsumer
import java.util.function.IntFunction
import java.lang.IndexOutOfBoundsException
import java.util.concurrent.atomic.AtomicReference
import net.minestom.server.utils.collection.MappedCollection
import org.jetbrains.annotations.ApiStatus
import java.lang.UnsupportedOperationException
import java.util.function.Function

@ApiStatus.Internal
class MappedCollection<O, R> : MutableCollection<R> {
    override fun size(): Int {
        return original.size
    }

    override fun isEmpty(): Boolean {
        return original.isEmpty()
    }

    override operator fun contains(o: Any): Boolean {
        for (entry in original) {
            if (mapper.apply(entry) == o) return true
        }
        return false
    }

    override fun iterator(): MutableIterator<R> {
        val iterator: Iterator<O> = original.iterator()
        return object : MutableIterator<R> {
            override fun hasNext(): Boolean {
                return iterator.hasNext()
            }

            override fun next(): R {
                return mapper.apply(iterator.next())
            }
        }
    }

    override fun toArray(): Array<Any> {
        // TODO
        throw UnsupportedOperationException("Unsupported array object")
    }

    override fun <T> toArray(a: Array<T>): Array<T> {
        // TODO
        throw UnsupportedOperationException("Unsupported array generic")
    }

    override fun containsAll(c: Collection<*>): Boolean {
        if (c.size > original.size) return false
        for (entry in c) {
            if (!contains(entry)) return false
        }
        return true
    }

    override fun add(t: R): Boolean {
        throw UnsupportedOperationException("Unmodifiable collection")
    }

    override fun remove(o: Any): Boolean {
        throw UnsupportedOperationException("Unmodifiable collection")
    }

    override fun addAll(c: Collection<R>): Boolean {
        throw UnsupportedOperationException("Unmodifiable collection")
    }

    override fun removeAll(c: Collection<*>): Boolean {
        throw UnsupportedOperationException("Unmodifiable collection")
    }

    override fun retainAll(c: Collection<*>): Boolean {
        throw UnsupportedOperationException("Unmodifiable collection")
    }

    override fun clear() {
        throw UnsupportedOperationException("Unmodifiable collection")
    }

    companion object {
        @JvmStatic
        fun <O : AtomicReference<R>?, R> plainReferences(original: Collection<O>): MappedCollection<O, R> {
            return MappedCollection(original, Function { obj: O -> obj!!.plain })
        }
    }
}